camayoc/tests/qcs/api/v1/scanjobs/test_run_scanjobs.py

```diff
@@ -303,3 +324,56 @@ def test_scan_task_results(scan_info):
         num_failed = task['systems_failed']
         num_scanned = task['systems_scanned']
         assert num_scanned == sys_count - num_failed
+
+
+@pytest.mark.parametrize(
+    'scan_info', scan_info(), ids=utils.name_getter)
+def test_products_found_deployment_report(scan_info):
+    """Test that the detail report contains expected data for the source.
+
+    :id: d5d424bb-8183-4b60-b21a-1b4ed1d879c0
+    :description: Test that the detail report contains expected data.
+    :steps:
+        1) Retreive the json report for the scan.
+        2) Assert that any products marked as present are expected to be found
+           as is listed in the configuration file for the source.
+    :expectedresults: There are inspection results for each source we scanned
+        and any products found are correctly identified.
+    """
+    result = get_scan_result(scan_info['name'])
+    report_id = result['report_id']
+    report = api.Client().get(
+        'reports/{}/deployments'.format(report_id)).json().get('report')
+    for entity in report:
+        present_products = []
+        all_found_products = []
+        for product in entity.get('products'):
+            name = ''.join(product['name'].lower().split())
+            if product['presence'] == 'present':
+                present_products.append(name)
+            if product['presence'] in ['present', 'potential']:
+                all_found_products.append(name)
+        for source_to_product_map in result['expected_products']:
+            src_id = list(source_to_product_map.keys())[0]
+            hostname = result['source_id_to_hostname'][src_id]
+            ex_products = source_to_product_map[src_id]
+            expected_product_names = [
+                ''.join(name.lower().split('-')) for name in ex_products.keys()
+                ]
+            if src_id in [s['id'] for s in entity['sources']]:
+                # We assert that products marked as present are expected
+                # We do not assert that products marked as potential must
+                # actually be on server
+                for prod_name in present_products:
+                    # Assert that products marked "present"
+                    # Are actually expected on machine
+                    if prod_name not in expected_product_names:
```

Comment:
you could use

```python
assert prod_name in expected, (
    'Found {found_product} but only expected to find '
    '{expected_products} on {host_found_on}'
    .format(
        found_product=prod_name,
        expected_products=expected_product_names,
        host_found_on=hostname,
    )
)
```

Reply:
Also with this logic it will stop on the first non-expected product, do you think it will be better to collect the unexpected products and raise the assertion error with all unexpected on the message?

Reply:
I think that when I do the next PR for the OS testing I would like to maybe work on generating separate tests for each of these. So most of this logic would go into a function that would generate the permutations for the tests. Then each host would get its own result. What do you think about that?

Reply:
I think that might be better to have one test per machine maybe and list everything that is wrong with that in case of failures. That is better to debug and target the problematic machine.